### PR TITLE
Add support for Kafka record headers

### PR DESCRIPTION
**Pull request by evan-stripe**
This adds support for specifying and receiving [Kafka record headers](https://cwiki.apache.org/confluence/display/KAFKA/KIP-82+-+Add+Record+Headers) (added in Kafka 0.11) to both the HTTP and gRPC interfaces. For gRPC headers are simply added as an additional field to the ProdRq and ConsRs structs - easy. For consuming over HTTP, they're added as an additional field - also easy.

For producing over HTTP, the HTTP server looks for any HTTP header with the prefix "X-Kafka-". It strips that prefix to get the Kafka record header key, and treats the HTTP header value as a base 64 encoded value.

I'm open to feedback on the interfaces here - a few things that I rejected but I could be wrong about:
- Base64 decoding HTTP headers. At least for my use case, I expect header keys to be human-specified and human-readable; I don't need to support non-ASCII keys.
- Returning record headers as HTTP headers over the HTTP consume API. This would be kind of cute, but I think ultimately somewhat silly (given that we already have a structured body format, which the publish API doesn't)
- Stripping the "Header" prefix from "HeaderKey" and "HeaderValue" in the proto spec. The existing fields seem to bias towards explicit/verbose, so I tried to match the existing style.

Record headers are only supported on Kafka version 0.11 or later, so if the `Version` config variable isn't set to something sufficiently new, Sarama will [silently ignore headers](https://github.com/Shopify/sarama/blob/3c763ff04e6daa57d4a4614e5bcd908f2527c989/produce_set.go#L77-L102). It might be desirable to return an error if the cluster configuration doesn't support producing messages with headers, but it doesn't seem like that's currently threaded through `producer.T` or `consumer.T`, so I'm not sure how to detect that case.

### CODE REVIEW

**Comment by evan-stripe** [comment]
Hmm. I'm having a hard time figuring out why the tests keep failing, but it seems like timeouts talking to zookeeper rather than anything that I've changed. Not sure what to do about that.

File: kafkapixy.proto [hunk 1]
```diff
@@ -115,6 +115,14 @@ service KafkaPixy {
     rpc GetTopicMetadata (GetTopicMetadataRq) returns (GetTopicMetadataRs) {}
 }
 
+message RecordHeader {
+    // Key in the header key-value pair
+    bytes header_key = 1;
```
Comment (horkhe) [inline kafkapixy.proto]: The `header_` prefix seems to be superfluous. Could you please remove it.

File: producer/producer_test.go [hunk 2]
```diff
@@ -57,7 +57,7 @@ func (s *ProducerSuite) TestProduce(c *C) {
 	offsetsBefore := s.kh.GetNewestOffsets("test.4")
 
 	// When
-	_, err := p.Produce("test.4", sarama.StringEncoder("1"), sarama.StringEncoder("Foo"))
+	_, err := p.Produce("test.4", sarama.StringEncoder("1"), sarama.StringEncoder("Foo"), nil)
```
Comment (horkhe) [inline producer/producer_test.go]: I would have added produce with headers test here.

**Comment by horkhe** [comment]
There is no consume with headers test in `service_http_test.go`. Preferably tests in `service_http_test.go` and `service_grpc_test.go` should be similar.

**Comment by horkhe** [comment]
I would prefer the produce API to fail if a header is passed, but the cluster does not support them. I think we can add a check [here](https://github.com/mailgun/kafka-pixy/pull/144/files#diff-3ac192716a1df66bbd1008d327af8792R175) and [here](https://github.com/mailgun/kafka-pixy/pull/144/files#diff-3ac192716a1df66bbd1008d327af8792R190) and return an error.

**Comment by horkhe** [comment]
@evan-stripe overall I like it. Just a few minor and easily fixable comments. Thank you for your contribution! @thrawn01 do you have any comments?

**Comment by evan-stripe** [comment]
Thanks for the feedback! I think I've incorporated all of your requests into the latest version of the patch, but let me know if I've missed anything.

File: proxy/proxy.go [hunk 3]
```diff
@@ -172,13 +173,18 @@ func (p *T) stopAdmin() {
 //
 // Errors usually indicate a catastrophic failure of the Kafka cluster, or
 // missing topic if there cluster is not configured to auto create topics.
-func (p *T) Produce(topic string, key, message sarama.Encoder) (*sarama.ProducerMessage, error) {
+func (p *T) Produce(topic string, key, message sarama.Encoder, headers []sarama.RecordHeader) (*sarama.ProducerMessage, error) {
+	if len(headers) > 0 &&
+		!p.cfg.Kafka.Version.IsAtLeast(sarama.V0_11_0_0) {
```
Comment (horkhe) [inline proxy/proxy.go]: Could you please keep it on the same line as `if`. There is no 80 chars per line limitation.

File: proxy/proxy.go [hunk 4]
```diff
@@ -187,13 +193,18 @@ func (p *T) Produce(topic string, key, message sarama.Encoder) (*sarama.Producer
 
 // AsyncProduce is an asynchronously counterpart of the `Produce` function.
 // Errors are silently ignored.
-func (p *T) AsyncProduce(topic string, key, message sarama.Encoder) {
+func (p *T) AsyncProduce(topic string, key, message sarama.Encoder, headers []sarama.RecordHeader) {
+	if len(headers) > 0 &&
+		!p.cfg.Kafka.Version.IsAtLeast(sarama.V0_11_0_0) {
```
Comment (horkhe) [inline proxy/proxy.go]: Also please make it on the same line as `if`.

File: proxy/proxy.go (hunk 3 above)
Comment (evan-stripe) [reply proxy/proxy.go]: No problem, fixed.

**Comment by horkhe** [comment]
Two very minor comments. Otherwise I am ready to merge. But before doing that I would like another pair of eyes primarily on the HTTP interface, the GRPC is solid. @thrawn01 please take a look.

**Comment by horkhe** [comment]
Ough, almost forgot, if you could update the README and CHANGELOG that would be great. But let's do that after @thrawn01 gives his go on the changes.

**Comment by horkhe** [comment]
Maybe we should say that headers are expected to be utf-8 strings and if a use case demands binary then it is the clients responsibility to base64 or whatever? Just thinking out loud, I am not against base64 encoding.

**Comment by evan-stripe** [comment]
I think Sarama may actually be wrong to expose the header name as a `[]byte` instead of a `string`. Both the Java API and the KIP use a string: https://kafka.apache.org/11/javadoc/org/apache/kafka/common/header/Header.html, and Kafka protocol strings are decoded as UTF-8.

So, I think we can definitely get away with using strings in the HTTP publish interface. I'm now wondering whether we should change the GRPC interfaces and the HTTP consumer interface to use strings as well. Thoughts?

**Comment by horkhe** [comment]
Given what you said about the Java implementation, I vote for headers to be strings all the way through.

**Comment by evan-stripe** [comment]
Alright, updated everything to be strings.

**Comment by thrawn01** [comment]
This is really awesome! I tested it out and added support for it in the CLI (See #147)

I ran into some usability issues though that might throw some.
* Kafka-pixy defaults to the V0.8.2.2 Kafka protocol unless you specify `KAFKA_VERSION` in the environment. Which means adding a header will fail unless `KAFKA_VERSION=0.11.0.2` or better.
* Because of `RecordHeader.encode()` and `RecordHeader.decode()` JSON marshalling of the `RecordHeader` struct will always encode the value. This threw me when I was un-marshalling values for the CLI. It's not a big deal, as the `Record` struct also has this issue.

**Comment by evan-stripe** [comment]
Hmm. Given that values are arbitrary byte arrays (not necessarily UTF-8), I'm not entirely sure how to make the types more ergonomic here; I think you have to base64-encode by default.

Totally agreed on the ergonomics of the version number, though - I ran into this when I was originally developing this feature. We hadn't bumped the version setting from the default, and the patch we're running internally didn't have the error message, and it was really challenging to debug. (It'd be _super_ nice if Sarama actually used the [ApiVersions](https://kafka.apache.org/protocol#The_Messages_ApiVersions) API to negotiate the protocol versions, instead of requiring them to be hardcoded.)

There is a version number option that you can put in the config file, though. One option would be to make the error message more verbose; maybe something like "headers are not supported with this version of kafka; update proxies.kafka.version in your config file or set $KAFKA_VERSION", but that's pretty wordy for a Go error message.

**Comment by horkhe** [comment]
* The support for the ApiVersions also needs to be implemented in the Kafka-Pixy consumer because we are not using the one from Sarama;
* The long message is ok IMHO. We already have [this](https://github.com/mailgun/kafka-pixy/blob/master/consumer/consumer.go#L23) after all.

**Comment by horkhe** [comment]
Closed by accident, sorry.

**Comment by horkhe** [comment]
Please rebase and I will merge.

**Comment by horkhe** [comment]
...And now that we agreed on API, please, update the README and CHANGELOG.

**Comment by evan-stripe** [comment]
Alright, I took a stab at updating the README and the CHANGELOG

**Comment by horkhe** [comment]
@evan-stripe looks like a test needs to be fixed:

**Comment by evan-stripe** [comment]
Bah that's what I get for making the error messages better 😛

Fix coming momentarily...

File: service/service_http_test.go [hunk 5]
```diff
@@ -196,6 +196,87 @@ func (s *ServiceHTTPSuite) TestUtf8Message(c *C) {
...
+	c.Check(rs.StatusCode, Equals, http.StatusBadRequest)
+
+	body := ParseJSONBody(c, rs).(map[string]interface{})
+	c.Check(body["error"], Equals, "Invalid base64 encoding for header: X-Kafka-Foo")
+}
+
+// Submitting headers when they're not supported by the Kafka version setting
+// should produce an error
+func (s *ServiceHTTPSuite) TestProduceHeadersUnsupported(c *C) {
+	if s.proxyCfg.Kafka.Version.IsAtLeast(sarama.V0_11_0_0) {
```
Comment (thrawn01) [inline service/service_http_test.go]: I think it should be 
```
if !s.proxyCfg.Kafka.Version.IsAtLeast(sarama.V0_11_0_0) {
```

Comment (evan-stripe) [reply service/service_http_test.go]: ...fascinating. Why is this passing on newer versions? Or should I just not ask 🙈

Comment (evan-stripe) [reply service/service_http_test.go]: Oh I guess it skips there which means we're just not getting any signal. Cool - pushing a fix now. Thanks for the pointer!

Comment (evan-stripe) [reply service/service_http_test.go]: Err wait, no that's actually correct - we _don't_ want to run the test on new versions. I think the actual issue is that gocheck's matcher doesn't behave the way I expected (I think it's implicitly matching on the whole string?)

Comment (thrawn01) [reply service/service_http_test.go]: ah, see I should have read the test. LOL

**Comment by evan-stripe** [comment]
:tada:

Thanks for all the feedback and help!

**Pull request merged**